File: framework/languages/node/e2e/AutomaticTurnDispatch/Server/Delay/Handlers/delay-handler.ts
import { Injectable } from '@nestjs/common';
import type { ZLinkMessageContext, ZLinkRequestHandler } from '@zlink-systems/framework';
import type { DelayRes, DelayReq } from '../../../Shared/messages';
import { EvidenceStore } from '../Support/evidence-store';

@Injectable()
export class DelayHandler implements ZLinkRequestHandler<DelayReq, DelayRes> {
  constructor(private readonly evidence: EvidenceStore) {}

  async handle(request: DelayReq, context: ZLinkMessageContext): Promise<DelayRes> {
    void context;
    this.evidence.add(`delay-started|rid=${this.evidence.rid}|request=${request.requestId}|marker=${request.marker}|delayMs=${request.delayMs}`);
    await new Promise((resolve) => setTimeout(resolve, request.delayMs));
    this.evidence.add(`delay-completed|rid=${this.evidence.rid}|request=${request.requestId}|marker=${request.marker}`);
    return {
      requestId: request.requestId,
      marker: request.marker,
      nodeRid: this.evidence.rid
    };
  }
}

File: framework/languages/dotnet/e2e/AutomaticTurnDispatch/Server/Delay/DelayHandler.cs
using AutomaticTurnDispatch.Shared;
using Zlink.Framework.Contracts.Handlers;

namespace AutomaticTurnDispatch.Server.Delay;

internal sealed class DelayHandler(NodeOptions node, EvidenceStore evidence)
    : IZLinkRequestHandler<DelayReq, DelayRes>
{
    public async ValueTask<DelayRes> HandleAsync(
        DelayReq request,
        IZLinkMessageContext context,
        CancellationToken cancellationToken)
    {
        _ = context;
        evidence.Add($"delay-started|rid={node.Rid}|request={request.RequestId}|marker={request.Marker}");
        await Task.Delay(TimeSpan.FromMilliseconds(request.DelayMs), cancellationToken);
        evidence.Add($"delay-completed|rid={node.Rid}|request={request.RequestId}|marker={request.Marker}");
        return new DelayRes(request.RequestId, request.Marker, node.Rid);
    }
}
